File: packages/prebid_mobile_sdk_gam/ios/prebid_mobile_sdk_gam/Sources/prebid_mobile_sdk_gam/PrebidMobileSdkGamPlugin.swift
import Flutter
import UIKit

/// Companion plugin that adds Google Ad Manager (GAM) rendering on top of the
/// core prebid_mobile_sdk plugin. Registers the GAM banner PlatformView factory
/// and the GAM interstitial method channel.
public class PrebidMobileSdkGamPlugin: NSObject, FlutterPlugin {

    /// Retains the interstitial manager for the lifetime of the plugin.
    private static var interstitialManager: GamInterstitialManager?
    private static var rewardedManager: GamRewardedManager?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let factory = GamBannerAdViewFactory(messenger: registrar.messenger())
        registrar.register(factory, withId: "prebid_mobile_sdk_gam/banner")

        let nativeFactory = GamNativeAdViewFactory(messenger: registrar.messenger())
        registrar.register(nativeFactory, withId: "prebid_mobile_sdk_gam/native")

        interstitialManager = GamInterstitialManager(messenger: registrar.messenger())
        rewardedManager = GamRewardedManager(messenger: registrar.messenger())
    }
}

File: packages/prebid_mobile_sdk_gam/android/src/main/kotlin/com/prebid/prebid_mobile_sdk_gam/PrebidMobileSdkGamPlugin.kt
package com.prebid.prebid_mobile_sdk_gam

import android.app.Activity
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding

/// Companion plugin that adds Google Ad Manager (GAM) rendering on top of the
/// core prebid_mobile_sdk plugin. Registers the GAM banner PlatformView factory
/// and the GAM interstitial method channel.
class PrebidMobileSdkGamPlugin : FlutterPlugin, ActivityAware {

    private var activity: Activity? = null
    private var interstitialManager: GamInterstitialManager? = null
    private var rewardedManager: GamRewardedManager? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        binding.platformViewRegistry.registerViewFactory(
            "prebid_mobile_sdk_gam/banner",
            GamBannerAdViewFactory(binding.binaryMessenger) { activity },
        )
        binding.platformViewRegistry.registerViewFactory(
            "prebid_mobile_sdk_gam/native",
            GamNativeAdViewFactory(binding.binaryMessenger),
        )
        interstitialManager = GamInterstitialManager(binding.binaryMessenger) { activity }
        rewardedManager = GamRewardedManager(binding.binaryMessenger) { activity }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        interstitialManager?.dispose()
        interstitialManager = null
        rewardedManager?.dispose()
        rewardedManager = null
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onDetachedFromActivity() {
        activity = null
    }
}
